src/i2c/src/i2c.rs

```diff
@@ -0,0 +1,548 @@
+const I2C_FUNC_SMBUS_BYTE_DATA: u64 =
+    I2C_FUNC_SMBUS_READ_BYTE_DATA | I2C_FUNC_SMBUS_WRITE_BYTE_DATA;
+const I2C_FUNC_SMBUS_WORD_DATA: u64 =
+    I2C_FUNC_SMBUS_READ_WORD_DATA | I2C_FUNC_SMBUS_WRITE_WORD_DATA;
+const I2C_FUNC_SMBUS_ALL: u64 =
+    I2C_FUNC_SMBUS_BYTE | I2C_FUNC_SMBUS_BYTE_DATA | I2C_FUNC_SMBUS_WORD_DATA;
+
+/// SMBUS protocol definitions
+/// SMBUS read or write markers
+const I2C_SMBUS_WRITE: u8 = 0;
+const I2C_SMBUS_READ: u8 = 1;
+
+/// SMBus transaction types (size parameter in the above functions)
+const I2C_SMBUS_QUICK: u32 = 0;
+const I2C_SMBUS_BYTE: u32 = 1;
+const I2C_SMBUS_BYTE_DATA: u32 = 2;
+const I2C_SMBUS_WORD_DATA: u32 = 3;
+
+/// As specified in SMBus standard
+const I2C_SMBUS_BLOCK_MAX: usize = 32;
+
+#[repr(C)]
+union I2cSmbusData {
+    byte: u8,
+    word: u16,
+    block: [u8; I2C_SMBUS_BLOCK_MAX + 2],
+}
+
+#[repr(C)]
+struct I2cSmbusIoctlData {
+    read_write: u8,
+    command: u8,
+    size: u32,
+    data: *mut I2cSmbusData,
+}
+
+/// I2C protocol definitions
+pub const I2C_M_RD: u16 = 0x0001; // read data, from slave to master
+
+#[derive(Copy, Clone)]
+#[repr(C)]
+struct I2cMsg {
+    addr: u16,
+    flags: u16,
+    len: u16,
+    buf: *mut u8,
+}
+
+#[repr(C)]
+struct I2cRdwrIoctlData {
+    msgs: *mut I2cMsg,
+    nmsgs: u32,
+}
+
+/// I2C backend definitions
+pub struct I2cReq {
+    pub addr: u16,
+    pub flags: u16,
+    pub len: u16,
+    pub buf: Vec<u8>,
+}
+
+/// I2C adapter and helpers
+#[derive(Debug)]
+struct I2cAdapter {
+    fd: File,
+    bus: u32,
+    smbus: bool,
+}
+
+impl I2cAdapter {
+    fn new(dev: String, dryrun: bool) -> Result<I2cAdapter> {
+        Ok(I2cAdapter {
+            bus: 0,
+            smbus: false,
+            fd: match dryrun {
+                true => OpenOptions::new()
+                    .read(true)
+                    .write(true)
+                    .create(true)
+                    .open("/tmp/i2cbus")?,
+                false => OpenOptions::new().read(true).write(true).open(dev)?,
+            },
+        })
+    }
+
+    /// Sets client's address for an I2C adapter.
+    fn set_client_addr(&self, addr: usize, dryrun: bool) -> Result<()> {
+        if dryrun {
+            return Ok(());
+        }
+
+        let ret = unsafe { ioctl(self.fd.as_raw_fd(), I2C_SLAVE, addr as c_ulong) };
+
+        if ret == -1 {
+            println!("Failed to set client addr to {:x}", addr);
+            errno_result()
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Gets adapter's functionality
+    fn get_func(&mut self, dryrun: bool) -> Result<()> {
+        let func: u64 = I2C_FUNC_SMBUS_ALL;
+
+        if !dryrun {
+            let ret = unsafe { ioctl(self.fd.as_raw_fd(), I2C_FUNCS, &func) };
+
+            if ret == -1 {
+                println!("Failed to get I2C function");
+                return errno_result();
+            }
+        }
+
+        if (func & I2C_FUNC_I2C) != 0 {
```

Comment:
I2C_FUNC_SMBUS_ALL is defined to be a constant, so does "func".  Then only "func & I2C_FUNC_SMBUS_ALL" will get matched, what's the purpose for other branches?

Reply:
> I2C_FUNC_SMBUS_ALL is defined to be a constant, so does "func". Then only "func & I2C_FUNC_SMBUS_ALL" will get matched, what's the purpose for other branches?

I get what you are saying, that the compiler may optimize it, but it shouldn't as we are using the func variable as reference in unsafe rust. So its value get changed by the ioctl.

I have tried this earlier for both SMBUS and I2C transfers, and both work. I tried it again now by marking func as 0 by default, and I was able to get SMBUS working. So the compiler doesn't optimize it because of its usage in unsafe block.